Create a page on read preference, write concern, and read concern (#819)
mongodb/docs-node: source/fundamentals/crud/read-write-pref.txt

.. _node-crud-write-read-pref:

===============================================
Specify How CRUD Operations Run on Replica Sets
===============================================

.. facet::
   :name: genre
   :values: reference

.. meta:: 
   :keywords: node.js, customize, preferences, replica set, consistency

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the **write concern**, **read concern**, and
**read preference** configurations to modify the way that MongoDB runs
create, read, update, and delete (CRUD) operations on replica sets.

You can set write concern, read concern, and read preference options at the following
levels:

> - Client, which sets the *default for all operation executions* unless overridden

Author: Following the API chain, the inheritance change goes like this:
Operation > Command Operation > Aggregate
And as far as I know, operation inherits from the client, so that's that I'm going to put in, but flag for review in the tech review.

Server manual. For more detailed API descriptions, see the `WriteConcern API documentation
<{+api+}/classes/WriteConcern.html>`__.

A ``WriteConcern`` accepts the following parameters:

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Parameter
     - Description
  
   * - ``w`` *(optional)*
     - | Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
       | **Type:** `W <{+api+}/types/W.html>`__

   *  - ``wtimeoutMS`` *(optional)*
      - | Specifies a time limit to prevent write operations from blocking indefinitely 
        | **Type:** number

   * - ``journal`` *(optional)*
     - | Requests acknowledgment that the write operation has been written to the on-disk journal 
       | **Type:** boolean

Example: Set the Write Concern for a Single Write Operation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code uses custom ``WriteConcern`` settings while creating new a document:

.. code-block:: js

   myDB.myCollection.insertOne( 
      { name: "anotherDocumentName" }, 
      { writeConcern: 
        { w: 2, wtimeoutMS: 5000 } 
      } 
   );

> Example: Create and Use a New Write Concern using the fromOptions and apply Methods

Reviewer: Suggestion:

I think it could be helpful to make this title more concise to make it easier to scan the table of contents (see attached image).

I think "a New Write Concern" could be slightly misleading since that describes the instances rather than the information in the setting and that mentioning the method names isn't necessary.

Perhaps "Retrieve and Apply an Existing Write Concern" would convey the actions of the example. 

<img width="514" alt="Screenshot 2023-11-30 at 1 12 29 PM" src="https://github.com/mongodb/docs-node/assets/52428683/aa270192-a4e9-4f85-ab80-26957b926f75">

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   

> This code use the ``fromOptions`` method to construct a ``WriteConcern`` from the options of an

Reviewer: Suggestion:

Method notation from a [prior suggestion](https://github.com/mongodb/docs-node/pull/819#discussion_r1401153101) and a grammar fix:

```suggestion
This code uses the ``fromOptions()`` method to construct a ``WriteConcern`` from the options of an
```

existing database, ``myDB``. Note that ``myDB`` could be replaced with any entity that

> includes a write concern option.

Reviewer: ```suggestion
accepts a write concern option.
```

.. code-block:: js

   const newWriteConcern = WriteConcern.fromOptions(myDB);
   const myDoc = { name: "New Document" };
   WriteConcern.apply(myDoc,newWriteConcern);

Read Concern
------------

The read concern specifies the following behaviors:

> - Level of :manual:`consistency </core/causal-consistency-read-write-concerns/>` across replica sets

Reviewer: Suggestion:

Sorry that I didn't mention this in a previous suggestion, but I think the complete term name as well as the section to which this links always uses  "causal consistency", which is a type of consistency. This could help disambiguate it from other types of consistency such as [data consistency](https://en.wikipedia.org/wiki/Data_consistency).

For more information about read concerns or read concern levels, see 
:manual:`Read Concern </reference/read-concern/>` in the Server manual. For more detail on
the ``ReadConcern`` type and definitions of the read concern levels, see the `ReadConcern <{+api+}/classes/ReadConcern.html>`__ in
the API documentation.

Example: Set the Read Concern Level of an Aggregation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code sets the read concern level of an an aggregation to ``"majority"``:

.. code-block:: js

   const pipeline = [
      {"$match": {
        category: "KITCHENWARE",
      }},
      {"$unset": [
        "_id",
        "category",
      ]}
    ]; 
   
   result = await myDB.collection("mycollection")
      .aggregate(
        pipeline, 
        { readConcern: 
          { level: "available" } 
        } 
      );

For more information about aggregates, see the :ref:`nodejs-aggregation` page.
   
Example: Change the Read Concern of a Database
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code changes the read concern level of a database to ``"local"``:

.. code-block:: js

   const options = { readConcern: { level: "local" } };

>    const myDB = client.db("db-name", options);

Reviewer: Suggestion:

Just want to make sure that the [original suggestion](https://github.com/mongodb/docs-node/pull/819#discussion_r1401127153) was communicated clearly.

While both "db_name" and "db-name" are valid, I still think it could be good to use "mydb" or something similar (e.g. "myDB" used as an example on this manual page on Databases https://www.mongodb.com/docs/manual/core/databases-and-collections/#databases) since it would be consistent with the pattern used earlier in this doc for collection ("mycollection").

```suggestion
   const myDB = client.db("mydb", options);
```

create, read, update, and delete (CRUD) operations on replica sets.

You can set write concern, read concern, and read preference options at the following
levels:

- Client, which sets the *default for all operation executions* unless overridden
- Session
- Transaction
- Database
- Collection

These options allow you to customize the consistency and availability of the data in your
replica sets.

Write Concern
-------------

The write concern specifies the level of acknowledgement requested from MongoDB for write
operations, such as an insert or update, before the operation successfully returns.
Operations that do not specify an explicit write concern inherit the global default write
concern settings.

For more information, see :manual:`Write Concern </reference/write-concern/>` in the
Server manual. For detailed API documentation, see the `WriteConcern API documentation
<{+api+}/classes/WriteConcern.html>`__.

The following table describes the ``WriteConcern`` parameters:

.. list-table::
   :header-rows: 1
   :widths: 25 25 50

   * - Parameter
     - Type
     - Description
  
   * - ``w`` *(optional)*
     - `W <{+api+}/types/W.html>`__

>      - Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags

Reviewer: Question:
From looking at the linked API docs for "W", I only see the potential types are either a number or "majority". How do you specify tags?

>      - Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags

Author: either as a number or a the string, "majority" - I've used w as a variable in two examples, once as a number and once as a string - I think this avoid defining allowable values, similar to not including allowable for ReadConcern levels.

>      - Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags

Author: I changed it to "to mongod instances that are labelled specified tags"

.. code-block:: js

   myDB.myCollection.insertOne( 
      { name: "anotherDocumentName" }, 
      { writeConcern: 
        { w: 2, wtimeoutMS: 5000 } 
      } 
   );

Example: Retrieve and Apply an Existing Write Concern
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   
This code uses the ``fromOptions()`` method to construct a ``WriteConcern`` from the
options of an existing database, ``myDB``. Note that ``myDB`` could be replaced with any
entity that accepts a write concern option. Then the new write concern is applied to a
document, ``myDoc``.

.. code-block:: js

   const newWriteConcern = WriteConcern.fromOptions(myDB);
   const myDoc = { name: "New Document" };
   WriteConcern.apply(myDoc,newWriteConcern);

Read Concern
------------

The read concern specifies the following behaviors:

- Level of :manual:`causal consistency </core/causal-consistency-read-write-concerns/>` across replica sets

- `Isolation guarantees
  <https://www.mongodb.com/docs/current/core/read-isolation-consistency-recency/>`__ maintained during a query 

Lower consistency requirements may reduce latency. The default read concern level is

> ``local``. This means that the client returns the data from the replica set member that the

Reviewer: Suggestion:

I think it could be helpful to identify that the setting parameter is called "level" prior to the example. The Read Preference and Write Concern sections both include the parameter name (in a table).

"You can specify the read concern setting in the \`\`level\`\` parameter. The default..."

.. code-block:: js

   const pipeline = [
      {"$match": {
        category: "KITCHENWARE",
      }},
      {"$unset": [
        "_id",
        "category",
      ]}
    ]; 
   
   result = await myDB.collection("mycollection")
      .aggregate(
        pipeline, 
        { readConcern: 
          { level: "available" } 
        } 
      );

For more information about aggregates, see the :ref:`nodejs-aggregation` page.
   
Example: Change the Read Concern of a Database
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code changes the read concern level of a database to ``"local"``:

.. code-block:: js

   const options = { readConcern: { level: "local" } };
   const myDB = client.db("mydb", options);

Read Preference
---------------

The read preference determines which member of a replica set MongoDB reads when running a
query. You can also customize how the server evaluates members. 

For more detailed API descriptions, see the `ReadPreference API
documentation <{+api+}/classes/ReadPreference.html>`__.

The following table describes the ``ReadPreference`` parameters:

.. list-table::
   :widths: 25 25 50
   :header-rows: 1

>    * - Mode

Reviewer: Suggestion:

Similar to prior comment on Write Concern.

```suggestion
   * - Parameter
```

     - Sets various options, including :manual:`hedge </core/read-preference-hedge-option/>`
       and :manual:`maxStalenessSeconds </core/read-preference-staleness/>`, that can be
       applied to your read preference.

Example: Set Read Preference and Concerns for a Transaction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code sets the read preference, read concern, and write concern for the operations in
a transaction:

.. code-block:: js

   const transactionOptions = {
      readPreference: "primary",
      readConcern: { level: "local" },
      writeConcern: { w: "majority" },
   };

   const session = client.startSession();
   session.startTransaction(transactionOptions);
   // ...
   await session.commitTransaction();
   await session.endSession();

For more information about transactions, see :ref:`nodejs-transactions`.

Example: Set the Read Preference of a Cluster in the Connection String
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code example creates a MongoClient that uses the "secondary" read preference mode
when performing queries on a cluster:

.. code-block:: js

>    const uri = "mongodb+srv://<user>:<password>@@<cluster-url>?readPreference=secondary&maxStalenessSeconds=120";

Reviewer 2: ```suggestion
   const uri = "mongodb+srv://<user>:<password>@<cluster-url>?readPreference=secondary&maxStalenessSeconds=120";
```

     - | request acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
       | **Type:** `W ***************** <{+api+}/types/W.html>`__

   *  - ``wtimeoutMS`` *(optional)*
      - | specify a time limit to prevent write operations from blocking indefinitely
        | **Type:** number

   * - ``journal`` *(optional)*
     - | request acknowledgment that the write operation has been written to the on-disk journal
       | **Type:** boolean

``apply`` Method *(static)*
````````````````````````````

This method defines or updates the ``WriteConcern`` of a command document.

.. code-block:: js

   apply(command, writeConcern): Document

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Parameter
     - Description

   * - ``command``
     - | **Type:** Document

   * - ``writeConcern``
     - | **Type:** WriteConcern

Update a ``WriteConcern`` for an existing document
``````````````````````````````````````````````````

This code updates the write concern of an existing document:

.. code-block:: js

   document = {}
   writeConcern = new WriteConcern(2, 30, true)

>    WriteConcern.apply(document, writeConcern)

Author: To be determined

       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-majority/#mongodb-readconcern-readconcern.-majority->`.

   * - ``'linearizable'``
     - | The query returns data that reflects all successful writes issued that had a write concern of ``majority`` and were acknowledged before the start of the read operation. 
       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-linearizable/#mongodb-readconcern-readconcern.-linearizable->`.

   * - ``'snapshot'``
     - | The query returns a complete copy of the data in a ``mongod`` instance at a specific point in time. Only available for operations within multi-document transactions.  
       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-snapshot/#mongodb-readconcern-readconcern.-snapshot->`.

Specify a ``ReadConcern`` for a collection
``````````````````````````````````````````

This code specifies the read concern level of an existing collection to ``'majority'``:

.. code-block:: js

   rc = new ReadConcern('majority')
   db.collection.find({}).options{ readConcern: rc }

Read Preference
---------------

> The read concern determines how queries are routed, that is which member of a replica set is read. The default mode is to read from the primary member of a replica set.

Reviewer: Suggestion: 
I think this is correctly identifying a [style guide rule](https://www.mongodb.com/docs/meta/style-guide/writing/use-that-which-correctly/). Normally, you'd use "which" to specify further detail and use "that" to specify a consequence (restrictive).

I think it could be made more concise and avoid the restrictive clause, e.g.:
"The read concern determines which members of a replica set are read when performing queries."

> The read concern determines how queries are routed, that is which member of a replica set is read. The default mode is to read from the primary member of a replica set.

Author: Originally I was using "i.e." to further describe the routing, but Snooty told me to change it to "that is", and then gave me this error. I like your suggestion tho, so I'll make that change. 

But in the case where I wanted to keep both phrasings, I think the comma is necessary.

.. _node-crud-write-read-pref:

===================================
Modify Execution of CRUD Operations
===================================

.. facet::
   :name: genre
   :values: reference

.. meta:: 
   :keywords: node.js, customize, preferences, replica set, consistency

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to modify the way that the {+driver-long+} executes create, read, update, and delete (CRUD) operations using **write concern**, **read concern**, and **read preference** configurations for replica sets.

You can set write concern, read concern, and read preference options at
the following levels:

> - Client level, which sets the *default for all operation executions*

Reviewer 3: q: how are each of these set when using the node driver?

  unless overridden
- Session level
- Transaction level
- Database level
- Collection level

These options allow you to customize the consistency and availability of the data in your replica sets.

Write Concern
-------------

A write concern describes the level of acknowledgement requested from MongoDB for write operations, such as an insert or update, before the operation is returned as successful. 
Operations which do not specify an explicit write concern inherit the global default write concern settings.

For more information, see :manual:`Write Concern </reference/write-concern/>`. For more detailed API descriptions, see `API WriteConcern <{+api+}/classes/WriteConcern.html>`__

> Constructor

Author: Removed constructors

> Constructor

Author: In this case, an object doesn't need to be instantiated. I've removed the constructor and clarified in the example description.

~~~~~~~~~~~

A ``WriteConcern`` can define all, some or none of its accepted parameters:

.. code-block:: js

   new WriteConcern(w, wtimeoutMS, journal): WriteConcern

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Parameter
     - Description
  

>    * - ``w`` *(optional)*
>      - | request acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
>        | **Type:** `W ***************** <{+api+}/types/W.html>`__
> 
>    *  - ``wtimeoutMS`` *(optional)*
>       - | specify a time limit to prevent write operations from blocking indefinitely
>         | **Type:** number
> 
>    * - ``journal`` *(optional)*
>      - | request acknowledgment that the write operation has been written to the on-disk journal
>        | **Type:** boolean

Reviewer 3: i: for consistency with other docs, i'd capitalize and use third-person verbs

```suggestion
   * - ``w`` *(optional)*
     - | Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
       | **Type:** `W ***************** <{+api+}/types/W.html>`__

   *  - ``wtimeoutMS`` *(optional)*
      - | Specifies a time limit to prevent write operations from blocking indefinitely
        | **Type:** number

   * - ``journal`` *(optional)*
     - | Requests acknowledgment that the write operation has been written to the on-disk journal
       | **Type:** boolean
```

``apply`` Method *(static)*
~~~~~~~~~~~~~~~~~~~~~~~~~~~

This method defines or updates the ``WriteConcern`` of a command document.

.. code-block:: js

   apply(command, writeConcern): Document

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Parameter
     - Description

   * - ``command``
     - | **Type:** Document

   * - ``writeConcern``
     - | **Type:** WriteConcern

Set a ``WriteConcern`` for a single write operation
````````````````````````````````````````````````````

This code uses a custom ``WriteConcern`` while creating new a document:

.. code-block:: js

   db.myCollection.insertOne( 
      {
        name: 'anotherDocumentName'
      }, 
      {
        writeConcern: { w: 2, wtimeout: 5000 }
      } 
   )

.. tip::
   
   Use the ``fromOptions`` method to construct a ``WriteConcern`` from an existing set of options or ``WriteConcern``:

   .. code-block:: js

>       writeConcern2 = WriteConcern.fromOptions(writeConcern1)

Author: I rewrote the example to extract a write concern from an existing entity - probably a likely use case?

     - | The query returns the instance’s most recent data acknowledged as having been written to a majority of members in the replica set.  
       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-majority/#mongodb-readconcern-readconcern.-majority->`.

   * - ``'linearizable'``
     - | The query returns data that reflects all successful writes issued that had a write concern of ``majority`` and were acknowledged before the start of the read operation. 
       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-linearizable/#mongodb-readconcern-readconcern.-linearizable->`.

   * - ``'snapshot'``
     - | The query returns a complete copy of the data in a ``mongod`` instance at a specific point in time. Only available for operations within multi-document transactions.  
       | For more information, see the :rapid:`Read Concern specification </reference/read-concern-snapshot/#mongodb-readconcern-readconcern.-snapshot->`.

Set a ``ReadConcern`` of a cluster
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code sets the ``ReadConcern`` level of an existing cluster to ``'majority'``:

.. code-block:: js

>    atlas clusters advancedSettings update myCluster --readConcern 'majority'

Author: This had been replaced with a relevant example.

Server manual. For more detailed API descriptions, see the `WriteConcern API documentation
<{+api+}/classes/WriteConcern.html>`__.

A ``WriteConcern`` accepts the following parameters:

.. list-table::
   :header-rows: 1
   :widths: 25 75

   * - Parameter
     - Description
  
   * - ``w`` *(optional)*
     - | Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
       | **Type:** `W <{+api+}/types/W.html>`__

   *  - ``wtimeoutMS`` *(optional)*
      - | Specifies a time limit to prevent write operations from blocking indefinitely 
        | **Type:** number

   * - ``journal`` *(optional)*
     - | Requests acknowledgment that the write operation has been written to the on-disk journal 
       | **Type:** boolean

Example: Set the Write Concern for a Single Write Operation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code uses custom ``WriteConcern`` settings while creating new a document:

.. code-block:: js

   myDB.myCollection.insertOne( 
      {
        name: "anotherDocumentName"
      }, {
        writeConcern: { w: 2, wtimeoutMS: 5000 }
      } 

>    );

Author: I collapsed most of them, but put the `{ w: 2, wtimeoutMS: 5000 }` bracket as it's own line, to emphasis that that it's nested within the write concern. I personally find it easier to read recognize how things are nested this way.

  guarantees maintained during a query 

Lower consistency requirements may reduce latency. The default read concern level is
``local``. This means that the client returns the data from the replica set member that the
client is connected to, with no guarantee that the data has been written to all replica set members.

For more information about read concerns or read concern levels, see 
:manual:`Read Concern </reference/read-concern/>` in the Server manual. For more detail on
the ``ReadConcern`` type and definitions of the read concern levels, see the `ReadConcern <{+api+}/classes/ReadConcern.html>`__ in
the API documentation.

Example: Set the Read Concern Level of an Aggregation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code sets the read concern level of an an aggregation to ``"majority"``:

.. code-block:: js

   const pipeline = [
      {"$match": {
        category: "KITCHENWARE",
      }},
      {"$unset": [
        "_id",
        "category",
      ]}
    ]; 
   
   result = await myDB.collection("mycollection")
      .aggregate(
        pipeline, 
        { readConcern: {
          level: "available" 

>           } 

Author: Similar to above, I condensed this but put `{ level: "available" }` on it's own line to emphasize the nesting.

        } 
      );

For more information about aggregates, see the :ref:`nodejs-aggregation` page.
   
Example: Change the Read Concern of a Database
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code changes the read concern level of a database to ``"local"``:

.. code-block:: js

   const options = { readConcern: { level: "local" } };
   const myDB = client.db("db_name", options);

Read Preference
---------------

The read concern determines which member of a replica set MongoDB reads when executing a
query. The default mode, ``primary``, specifies that operations are read from the primary member of the replica set.

Read Preference Options
~~~~~~~~~~~~~~~~~~~~~~~

``mode``

> `````````

Author: Since I've removed the options descriptions, I've put the parameters into a table, so this no longer a heading.

     - Most operations read from a secondary member of the replica set. If a
       :ref:`read_preference_options` value is defined, it will be used to determine an
       appropriate member. If none of the secondary members meet the staleness criteria,
       the operation reads from the primary.

   * - ``"nearest"``
     - Operations read from a member that has a latency within the acceptable range.
       Whether a member is primary or secondary is not considered. We recommend setting a
       :ref:`read_preference_options` value to avoid returning stale data.

``tags`` *(optional)*
`````````````````````

A list of tags used to prioritize secondary members by using tags associated with each
member. Tags cannot be used with  ``"primary"`` mode. For more information about tag sets
or managing member tags, see :manual:`Read Preference Tag Set Lists
</core/read-preference-tags/#read-preference-tag-set-lists/>`.

.. _read_preference_options:

``options`` *(optional)*
````````````````````````

The following options may be sent as a `ReadPreferenceOptions
<{+api+}/interfaces/ReadPreferenceOptions.html>`__ object:

.. list-table::
   :widths: 25 75
   :header-rows: 1

>    * - Mode

Author: Removed.

     - | Requests acknowledgment that the write operation has propagated to a specified number of mongod instances or to mongod instances with specified tags 
       | **Type:** `W <{+api+}/types/W.html>`__

   *  - ``wtimeoutMS`` *(optional)*
      - | Specifies a time limit to prevent write operations from blocking indefinitely 
        | **Type:** number

   * - ``journal`` *(optional)*
     - | Requests acknowledgment that the write operation has been written to the on-disk journal 
       | **Type:** boolean

Example: Set the Write Concern for a Single Write Operation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This code uses custom ``WriteConcern`` settings while creating new a document:

.. code-block:: js

   db.myCollection.insertOne( 
      {
        name: 'anotherDocumentName'
      }, {
        writeConcern: { w: 2, wtimeoutMS: 5000 }
      } 
   )

Example: Create a New ``WriteConcern`` from an Existing Database's ``WriteConcern``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   
Use the ``fromOptions`` method to construct a ``WriteConcern`` from an existing entity's
options:

.. code-block:: js

>   const newWriteConcern = WriteConcern.fromOptions(db);

Author: I decided to use the `apply` method as well in this example, since I've already shown how to send a write concern using `insertOne`.

Read Concern
------------

The read concern determines the level of `consistency
<https://www.mongodb.com/docs/current/core/causal-consistency-read-write-concerns/>`__
across replica sets and the `isolation
<https://www.mongodb.com/docs/current/core/read-isolation-consistency-recency/>`__

> guarantees maintained during a query. You can specify your read concern level by

Author: I disagree that they aren't parallel, but I do think it's clearer to put them in a list.

constructing an instance of a ``ReadConcern`` or by specifying the options in an object. 

Lower consistency requirements may reduce latency. The default read concern level is
``local``. This means that the client returns the data from the local instance, with no
guarantee that the data has been written to all replica set members.

For more information about read concerns or read concern levels, see 

> :manual:`Read Concern </reference/read-concern/>` in the server manual. For more detailed 

Author: I've updated this, and I'll follow up with a PR to update the other instances in this repo that use the lower case in this situation.
